pygmt/gridops.py

```diff
@@ -0,0 +1,119 @@
+    {J}
+    {R}
+    extend : bool or int or float
+        Allow grid to be extended if new *region* exceeds existing boundaries.
+        Give a value to initialize nodes outside current region.
+    circ_subregion : str
+        ``'lon/lat/radius[unit][+n]'``.
+        Specify an origin (*lon* and *lat*) and *radius*; append a distance
+        *unit* and we determine the corresponding rectangular region so that
+        all grid nodes on or inside the circle are contained in the subset.
+        If **+n** is appended we set all nodes outside the circle to NaN.
+    z_subregion : str
+        ``'[min/max][+n|N|r]'``.
+        Determine a new rectangular region so that all nodes outside this
+        region are also outside the given z-range [-inf/+inf]. To indicate no
+        limit on *min* or *max* only, specify a hyphen (-). Normally, any NaNs
+        encountered are simply skipped and not considered in the
+        range-decision. Append **+n** to consider a NaN to be outside the given
+        z-range. This means the new subset will be NaN-free. Alternatively,
+        append **+r** to consider NaNs to be within the data range. In this
+        case we stop shrinking the boundaries once a NaN is found [Default
+        simply skips NaNs when making the range decision]. Finally, if your
+        core subset grid is surrounded by rows and/or columns that are all
+        NaNs, append **+N** to strip off such columns before (optionally)
+        considering the range of the core subset for further reduction of the
+        area.
+
+    Returns
+    -------
+    ret: xarray.DataArray or None
+        Return type depends on whether the *outgrid* parameter is set:
+
+        - xarray.DataArray if *outgrid* is not set
+        - None if *outgrid* is set (grid output will be stored in *outgrid*)
+    """
+    kind = data_kind(grid)
+
+    with GMTTempFile(suffix=".nc") as tmpfile:
+        with Session() as lib:
+            if kind == "file":
+                file_context = dummy_context(grid)
+            elif kind == "grid":
+                raise NotImplementedError(
+                    "xarray.DataArray is not supported as the input grid yet!"
+                )
+                # file_context = lib.virtualfile_from_grid(grid)
+                # See https://github.com/GenericMappingTools/gmt/pull/3532
+                # for a feature request.
+            else:
+                raise GMTInvalidInput("Unrecognized data type: {}".format(type(grid)))
+
+            with file_context as infile:
+                if "G" not in kwargs.keys():  # if outgrid is unset, output to tempfile
+                    kwargs.update({"G": tmpfile.name})
+                outgrid = kwargs["G"]
+                arg_str = " ".join([infile, build_arg_string(kwargs)])
+                lib.call_module("grdcut", arg_str)
+
+        if outgrid == tmpfile.name:  # if user did not set outgrid, return DataArray
+            with xr.open_dataarray(outgrid) as dataarray:
```

Comment:
After starting #494, I'm starting to wonder if we should have a `pygmt.open` function to load NetCDF files **with** the `node_offset` attribute inside the xarray.DataArray, as needed if we're going to start chaining operations as you mentioned in https://github.com/GenericMappingTools/pygmt/pull/476#issuecomment-649872282.

Reply:
If I understand the GMT C API correctly, GMT has its own data types (internally) for [grids](https://docs.generic-mapping-tools.org/dev/api.html#gmt-grids), [CPT](https://docs.generic-mapping-tools.org/dev/api.html#color-palette-tables-cpt) and [data tables](https://docs.generic-mapping-tools.org/dev/api.html#data-tables). It also provides API functions [`GMT_Read_Data`](https://docs.generic-mapping-tools.org/dev/api.html#gmt-read-data) and [`GMT_Write_Data`](https://docs.generic-mapping-tools.org/dev/api.html#gmt-write-data), and two special modules [`gmtread`](https://github.com/GenericMappingTools/gmt/blob/master/src/gmtread.c) and [`gmtwrite`](https://github.com/GenericMappingTools/gmt/blob/master/src/gmtwrite.c) which I presume can read and write any GMT data types. Perhaps in the future we need to wrap these GMT data types. In this case, `pygmt.read()` and `pygmt.write()` may make more sense.